galaxy_ng/app/api/v3/viewsets/collection.py

```diff
@@ -87,21 +85,75 @@ class CollectionViewSet(api_base.LocalSettingsMixin,
-                                          pulp_ansible_views.UnpaginatedCollectionVersionViewSet):
+class UnpaginatedCollectionVersionViewSet(
+    api_base.LocalSettingsMixin,
+    ViewNamespaceSerializerContextMixin,
+    pulp_ansible_views.UnpaginatedCollectionVersionViewSet,
+):
     pagination_class = None
     serializer_class = UnpaginatedCollectionVersionSerializer
     permission_classes = [access_policy.CollectionAccessPolicy]
 
 
+def get_dependents(parent):
+    """Given a parent collection version, return a list of
+    collection versions that depend on it.
+    """
+    key = f"{parent.namespace}.{parent.name}"
+    dependents = []
+    for child in CollectionVersion.objects.filter(dependencies__has_key=key):
+        spec = SimpleSpec(child.dependencies[key])
+        if spec.match(Version(parent.version)):
+            dependents.append(child)
+    return dependents
+
+
 class CollectionVersionViewSet(api_base.LocalSettingsMixin,
                                ViewNamespaceSerializerContextMixin,
                                pulp_ansible_views.CollectionVersionViewSet):
     serializer_class = CollectionVersionSerializer
     permission_classes = [access_policy.CollectionAccessPolicy]
     list_serializer_class = CollectionVersionListSerializer
 
+    def destroy(self, request: Request, *args, **kwargs) -> Response:
+        """
+        Allow a CollectionVersion to be deleted.
+        1. Perform Dependency Check to verify that the collection version can be deleted
+        2. If the collection version can’t be deleted, return the reason why
+        3. If it can, dispatch task to delete CollectionVersion and clean up repository.
+           If the version being deleted is the last collection version in the collection,
+           remove the collection object as well.
+        """
+        collection_version = self.get_object()
+
+        # dependency check
+        dependents = get_dependents(collection_version)
+        if dependents:
+            return Response(
+                {
+                    "detail": _(
+                        "Collection version {namespace}.{name} {version} could not be "
+                        "deleted because there are other collections that require it."
+                    ).format(
+                        namespace=collection_version.namespace,
+                        name=collection_version.collection.name,
+                        version=collection_version.version,
+                    ),
+                    "dependent_collection_versions": [
+                        f"{dep.namespace}.{dep.name} {dep.version}" for dep in dependents
+                    ],
+                },
+                status=400,
+            )
+
+        dispatch(
+            delete_collection_version,
+            exclusive_resources=collection_version.repositories.all(),
+            kwargs={"collection_version_pk": collection_version.pk},
+        )
```

Comment:
At the top of the file

```py
from pulpcore.plugin.viewsets import OperationPostponedResponse
```

then

```suggestion
        async_result = dispatch(
            delete_collection_version,
            exclusive_resources=collection_version.repositories.all(),
            kwargs={"collection_version_pk": collection_version.pk},
        )
        return OperationPostponedResponse(async_result, request)
```

This makes the proper format with the task link on the response, I just learned this, I will update the namespace deletion properly, we have to do that on every endpoint where a task is dispatched, so the response is always the link to the task (clickable link on API ui)